lang/wasm_instance_cache.hh

```diff
@@ -0,0 +1,326 @@
+                            sm::description("The number of user defined functions loaded")),
+            sm::make_counter("cache_blocks", wasm::instance_cache::shard_stats().cache_blocks,
+                            sm::description("The number of times a user defined function waited for an instance")),
+            sm::make_gauge("cache_total_size", [this] { return _total_size; },
+                            sm::description("The total size of instances stored in the user defined function cache")),
+        });
+    }
+private:
+    // key with a copied function name to store in the cache
+    struct cache_key_type {
+        seastar::scheduling_group scheduling_group;
+        cql3::functions::function_name name;
+    public:
+        cache_key_type(seastar::scheduling_group scheduling_group, const cql3::functions::function_name& name)
+            : scheduling_group(std::move(scheduling_group))
+            , name(name) {
+        }
+        bool operator==(const cache_key_type& other) const {
+            return scheduling_group == other.scheduling_group && name == other.name;
+        }
+    };
+    // temporary key with a reference to the function name to find elements in the cache
+    struct lookup_cache_key_type {
+        seastar::scheduling_group scheduling_group;
+        const cql3::functions::function_name& name;
+    public:
+        lookup_cache_key_type(seastar::scheduling_group scheduling_group, const cql3::functions::function_name& name) noexcept
+            : scheduling_group(std::move(scheduling_group))
+            , name(name) {
+        }
+        explicit lookup_cache_key_type(const cache_key_type& key) noexcept
+            : scheduling_group(key.scheduling_group)
+            , name(key.name) {
+        }
+        bool operator==(const cache_key_type& other) const {
+            return scheduling_group == other.scheduling_group && name == other.name;
+        }
+    };
+
+    struct cache_key_hash {
+        using is_transparent = void;
+
+        size_t operator()(const cache_key_type& v) const noexcept {
+            return utils::tuple_hash()(v.scheduling_group, v.name);
+        }
+        size_t operator()(const lookup_cache_key_type& v) const noexcept {
+            return utils::tuple_hash()(v.scheduling_group, v.name);
+        }
+    };
+
+    struct lru_entry_type {
+        lookup_cache_key_type key;
+        seastar::lowres_clock::time_point timestamp;
+        size_t instance_size;
+        lru_entry_type(const cache_key_type& key, seastar::lowres_clock::time_point timestamp, size_t size) noexcept :
+            key(key), timestamp(timestamp), instance_size(size) {}
+    };
+
+    struct cache_entry_type {
+        seastar::shared_mutex mutex;
+        seastar::lw_shared_ptr<wasm_instance> instance;
+        // iterator is invalid when the entry is being used (at that point, it is not in lru)
+        std::list<lru_entry_type>::iterator it;
+    };
```

Comment:
I see there is some cooperation between users of the instance and the cache, since the users need to understand whether or not to release the entry. We can hide this knowledge leak by having a run_script() method for the cache entry and exposing it. Not sure it's worth it at this stage.